Mathlib/NumberTheory/ModularForms/BoundedAtCusp.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2025 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.Analysis.Complex.UpperHalfPlane.FunctionsBoundedAtInfty
+import Mathlib.NumberTheory.ModularForms.Cusps
+import Mathlib.NumberTheory.ModularForms.SlashActions
+
+/-!
+# Boundedness and vanishing at cusps
+
+We define the notions of "bounded at c" and "vanishing at c" for functions on `ℍ`, where `c` is
+an element of `OnePoint ℝ`.
+-/
+
+open Matrix SpecialLinearGroup UpperHalfPlane Filter Polynomial OnePoint
+
+open scoped MatrixGroups LinearAlgebra.Projectivization ModularForm
+
+namespace UpperHalfPlane
+
+variable {g : GL (Fin 2) ℝ} {f : ℍ → ℂ} (k : ℤ)
+
+lemma IsZeroAtImInfty.slash (hg : g 1 0 = 0) (hf : IsZeroAtImInfty f) :
+    IsZeroAtImInfty (f ∣[k] g) := by
+  rw [IsZeroAtImInfty, ZeroAtFilter, tendsto_zero_iff_norm_tendsto_zero] at hf ⊢
+  simpa [ModularForm.slash_def, denom, hg, mul_assoc]
+    using (hf.comp <| tendsto_smul_atImInfty hg).mul_const _
+
+lemma IsBoundedAtImInfty.slash (hg : g 1 0 = 0) (hf : IsBoundedAtImInfty f) :
+    IsBoundedAtImInfty (f ∣[k] g) := by
+  rw [IsBoundedAtImInfty, BoundedAtFilter, ← Asymptotics.isBigO_norm_left] at hf ⊢
+  suffices (fun x ↦ (‖g.det.val ^ (k - 1)‖ * ‖g 1 1 ^ (-k)‖) * ‖f (g • x)‖) =O[atImInfty] 1 by
+    simpa [ModularForm.slash_def, denom, hg, mul_assoc, mul_comm ‖f _‖]
+  apply (hf.comp_tendsto (tendsto_smul_atImInfty hg)).const_mul_left
+
+end UpperHalfPlane
+
+namespace OnePoint
+
+variable (c : OnePoint ℝ) (f : ℍ → ℂ) (k : ℤ)
+
+/-- We say `f` is bounded at `c` if, for all `g` with `g • ∞ = c`, the function `f ∣[k] g` is
+bounded at `∞`. -/
+def IsBoundedAt : Prop := ∀ g : GL (Fin 2) ℝ, g • ∞ = c → IsBoundedAtImInfty (f ∣[k] g)
+
+/-- We say `f` is zero at `c` if, for all `g` with `g • ∞ = c`, the function `f ∣[k] g` is
+zero at `∞`. -/
+def IsZeroAt : Prop := ∀ g : GL (Fin 2) ℝ, g • ∞ = c → IsZeroAtImInfty (f ∣[k] g)
+
+variable {c f k} {g : GL (Fin 2) ℝ}
+
+lemma IsBoundedAt.slash : IsBoundedAt c (f ∣[k] g) k ↔ IsBoundedAt (g • c) f k := by
```

Comment:
I am not sure about the name, maybe it should contain `iff`. You can add an alias for one direction to allow not notation if you want.